### ISSUE TITLE
How is the +ms segmentation head implemented ?

### ISSUE THREAD

**Issue by sabeaussan** [facebookresearch/dinov2#394]
Hi all,

I'm trying to implement a multi-scale segmentation head as described in the paper : 

> "+ms: a boosted version of the linear setup. We concatenate the patch tokens of the 4 last layers, use a larger image resolution of 640, and use multiscale test-time augmentations to improve the predictions".

 I get that I should use the get_intermediate_layers() method to retrieve n last layers features, but what exactly are you referring to with "multiscale test-time augmentations" ? I found this snippet of code that seems to implement this part but I can't understand what is going on in this code :
`def _transform_inputs(self, inputs):
        """Transform inputs for decoder.
        Args:
            inputs (list[Tensor]): List of multi-level img features.
        Returns:
            Tensor: The transformed inputs
        """

        if self.input_transform == "resize_concat":
            # accept lists (for cls token)
            input_list = []
            for x in inputs:
                if isinstance(x, list):
                    input_list.extend(x)
                else:
                    input_list.append(x)
            inputs = input_list
            # an image descriptor can be a local descriptor with resolution 1x1
            for i, x in enumerate(inputs):
                if len(x.shape) == 2:
                    inputs[i] = x[:, :, None, None]
            # select indices
            inputs = [inputs[i] for i in self.in_index]
            # Resizing shenanigans
            # print("before", *(x.shape for x in inputs))
            if self.resize_factors is not None:
                assert len(self.resize_factors) == len(inputs), (len(self.resize_factors), len(inputs))
                inputs = [
                    resize(input=x, scale_factor=f, mode="bilinear" if f >= 1 else "area")
                    for x, f in zip(inputs, self.resize_factors)
                ]
                # print("after", *(x.shape for x in inputs))
            upsampled_inputs = [
                resize(input=x, size=inputs[0].shape[2:], mode="bilinear", align_corners=self.align_corners)
                for x in inputs
            ]
            inputs = torch.cat(upsampled_inputs, dim=1)`

The features are upscaled to resize_factors and then, the next line, down scaled to original shape. What's the point of doing that ?

**Comment by uebian** [comment]
I recently looked into this issue and came to the following understanding:

The `+ms` option comes from the `mmsegmentation` library. At test time, it enables multi-scale inference, where each input image is resized to several predefined scales (for DinoV2, these are `[1.0, 1.32, 1.73, 2.28, 3.0]`).  
These values are specified in the config file here:  
https://dl.fbaipublicfiles.com/dinov2/dinov2_vitb14/dinov2_vitb14_ade20k_ms_config.py  

```python
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(99999999, 640),
        img_ratios=[1.0, 1.32, 1.73, 2.28, 3.0],
        flip=True,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img'])
        ])
]
```

Each resized image is passed through the model independently. The resulting logits are then averaged before applying `argmax` to obtain the final segmentation map.
This logic is implemented in the `aug_test` method of `mmsegmentation`'s `0.x` branch (used by DinoV2) https://github.com/open-mmlab/mmsegmentation/blob/f67ef9c128eb2b643beaed8eb518c9fa09eb0912/mmseg/models/segmentors/encoder_decoder.py#L290:

```python
def aug_test(self, imgs, img_metas, rescale=True):
    """Test with augmentations.

    Only rescale=True is supported.
    """
    # aug_test rescales all imgs back to ori_shape
    assert rescale
    # to save memory, we accumulate augmented seg logits inplace
    seg_logit = self.inference(imgs[0], img_metas[0], rescale)
    for i in range(1, len(imgs)):
        cur_seg_logit = self.inference(imgs[i], img_metas[i], rescale)
        seg_logit += cur_seg_logit
    seg_logit /= len(imgs)

    if self.out_channels == 1:
        seg_pred = (seg_logit > self.decode_head.threshold).to(seg_logit).squeeze(1)
    else:
        seg_pred = seg_logit.argmax(dim=1)

    seg_pred = seg_pred.cpu().numpy()
    # unravel batch dimension
    seg_pred = list(seg_pred)
    return seg_pred
```
